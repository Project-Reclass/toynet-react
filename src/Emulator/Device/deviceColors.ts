/*
Copyright (C) 1992-2021 Free Software Foundation, Inc.

This file is part of ToyNet React.

ToyNet React is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 3, or (at your option) any later
version.

ToyNet React is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License
along with ToyNet React; see the file LICENSE.  If not see
<http://www.gnu.org/licenses/>.

*/
import { DeviceType } from 'src/common/types';

export type LinkFunc = (from: string, to: string) => any;

export enum DeviceColor {
  EMPTY = '#707070',
  ROUTER = '#BDA913',
  SWITCH = '#008A9E',
  HOST = '#9E1059',
}

export const deviceColorClasses = new Map<DeviceType | 'empty', DeviceColor>();
deviceColorClasses.set('router', DeviceColor.ROUTER);
deviceColorClasses.set('switch', DeviceColor.SWITCH);
deviceColorClasses.set('host', DeviceColor.HOST);
deviceColorClasses.set('empty', DeviceColor.EMPTY);