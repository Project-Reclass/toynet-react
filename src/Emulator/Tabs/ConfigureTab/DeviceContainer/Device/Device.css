.device-box {
  display: flex;
  width: 100%;
  background-color: #595D62;
  border-radius: 10px;
  padding: 5px;
  margin-bottom: 10px;
}

.device-name-box {
  text-transform: capitalize;
  display: flex;
  font-size: large;
  height: 50px;
  width: 60px;
  border-radius: 10px;
  justify-content: center;
  align-items: center;
  margin-right: 20px;
}

.vertical-bar {
  border-right: 2px solid #B2B2B2;
}

.connections-container {
  font-size: small;
  color: #B2B2B2;
  margin-left: 20px;
}

.router-color {
  background-color: #BDA913;
}

.switch-color {
  background-color: #9E1059;
}

.host-color {
  background-color: #008A9E;
}