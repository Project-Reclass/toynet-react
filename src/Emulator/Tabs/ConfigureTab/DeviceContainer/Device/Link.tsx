import React, { FC } from 'react';
import { useDrag } from 'react-dnd';

import { Connection } from './styled';

interface Props {
  connection: string;
}

const Link: FC<Props> = ({ connection }) => {
  const [, drag] = useDrag({
    item: {
      type: 'device',
      deviceData: {
        name: connection,
        connections: [],
        isLink: true,
      },
    },
  });

  return (
    <Connection
      isDragging={false}
      ref={drag}
      key={`${connection}`}
      id={`${connection}`}
      type={connection[0].toLocaleLowerCase()}
      style={{ transform: 'rotateX(-180deg)' }}
    >
      {connection}
    </Connection>
  );
};

export default Link;