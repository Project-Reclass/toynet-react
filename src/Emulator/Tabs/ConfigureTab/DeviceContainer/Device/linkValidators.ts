import { DeviceInterface } from 'src/common/types';

type LinkValidator = (from: DeviceInterface, to: DeviceInterface) => boolean;

const linkValidators = new Map<string, LinkValidator>();
linkValidators.set('s', validateSwitchLink);
linkValidators.set('h', validateHostLink);
linkValidators.set('r', validateRouterLink);

function validateSwitchLink(from: DeviceInterface, to: DeviceInterface) {
  if (to.name.startsWith('s') || to.name.startsWith('r') || (to.name.startsWith('h') && to.connections.length === 0))
    if (to.connections.indexOf(from.name) === -1)
      return true;

  return false;
};

function validateRouterLink(from: DeviceInterface, to: DeviceInterface) {
  if (to.name.startsWith('r') || to.name.startsWith('s'))
    if (to.connections.indexOf(from.name) === -1)
      return true;
  return false;
};

function validateHostLink(from: DeviceInterface, to: DeviceInterface) {
  if (to.name.startsWith('s') && from.connections.length === 0)
    if (to.connections.indexOf(from.name) === -1)
      return true;
  return false;
};

export function isValidLink(from: DeviceInterface, to: DeviceInterface) {
  if (from.name.length < 1 || to.name.length < 1 || from.name === to.name)
    return false;
  const validator = linkValidators.get(from.name[0]);
  return validator ? validator(from, to) : false;
};

export default isValidLink;