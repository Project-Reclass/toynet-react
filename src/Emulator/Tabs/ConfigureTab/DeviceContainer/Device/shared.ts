export type LinkFunc = (from: string, to: string) => any;

export enum DeviceColor {
  EMPTY = '#707070',
  ROUTER = '#BDA913',
  SWITCH = '#008A9E',
  HOST = '#9E1059',
}

export const deviceColorClasses = new Map<string, DeviceColor>();
deviceColorClasses.set('Router', DeviceColor.ROUTER);
deviceColorClasses.set('r', DeviceColor.ROUTER);
deviceColorClasses.set('Switch', DeviceColor.SWITCH);
deviceColorClasses.set('s', DeviceColor.SWITCH);
deviceColorClasses.set('Host', DeviceColor.HOST);
deviceColorClasses.set('h', DeviceColor.HOST);
deviceColorClasses.set('empty', DeviceColor.EMPTY);