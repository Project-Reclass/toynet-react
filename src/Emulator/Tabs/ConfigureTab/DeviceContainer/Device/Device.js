import React from "react";
import "./Device.css";

const Device = ({ deviceName, deviceData }) => {
  let boxClass = "device-name-box ";
  if (deviceName === "Router") {
    boxClass += "router-color";
  } else if (deviceName === "Switch") {
    boxClass += "switch-color";
  } else {
    boxClass += "host-color";
  }

  return (
    <div className="device-box">
      <div className={boxClass}>
        <div>{deviceData.name}</div>
      </div>
      <div> --&gt; </div>
      {deviceData.connections}
    </div>
  );
};

export default Device;