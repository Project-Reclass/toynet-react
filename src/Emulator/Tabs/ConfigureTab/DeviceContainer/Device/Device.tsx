import React, { FC, useMemo } from 'react';
import { useDrag, useDrop } from 'react-dnd';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTrashAlt } from '@fortawesome/free-solid-svg-icons';

import { mergeRefs } from 'src/common/utils';
import { DeviceInterface } from 'src/common/types';

import './Device.css';

enum DeviceColor {
  EMPTY = 'empty-color',
  ROUTER = 'router-color',
  SWITCH = 'switch-color',
  HOST = 'host-color',
}

const deviceColorClasses = new Map<string, DeviceColor>();
deviceColorClasses.set('Router', DeviceColor.ROUTER);
deviceColorClasses.set('r', DeviceColor.ROUTER);
deviceColorClasses.set('Switch', DeviceColor.SWITCH);
deviceColorClasses.set('s', DeviceColor.SWITCH);
deviceColorClasses.set('Host', DeviceColor.HOST);
deviceColorClasses.set('h', DeviceColor.HOST);

export function mergeRefs<T = any>(
  refs: Array<React.MutableRefObject<T> | React.LegacyRef<T>>,
): React.RefCallback<T> {
  return (value) => {
    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(value);
      } else if (ref != null) {
        (ref as React.MutableRefObject<T | null>).current = value;
      }
    });
  };
}

interface Props {
  deviceName: string;
  deviceData: DeviceInterface;
  onDrop: (from: string, to: string) => any;
}

type LinkValidator = (from: string, to: string, connections: string[]) => boolean;

const validateSwitchLink = (from: string, to: string, connections: string[]) => {
  if (from.startsWith('s') && (to.startsWith('s') || to.startsWith('r') || to.startsWith('h')))
    if (connections.indexOf(from) < 0)
      return true;
  return false;
};

const validateRouterLink = (from: string, to: string, connections: string[]) => {
  if (from.startsWith('r') && (to.startsWith('r') || to.startsWith('s')))
    if (connections.indexOf(from) < 0)
      return true;
  return false;
};

const validateHostLink = (from: string, to: string, connections: string[]) => {
  if (from.startsWith('h') && to.startsWith('s'))
    return true;
  return false;
};

const linkValidators = new Map<string, LinkValidator>();
linkValidators.set('s', validateSwitchLink);
linkValidators.set('h', validateHostLink);
linkValidators.set('r', validateRouterLink);

const isValidLink = (from: string, to: string, connections: string[]) => {
  if (from.length < 1 || to.length < 1 || from === to)
    return false;
  const validator = linkValidators.get(from[0]);
  return validator ? validator(from, to, connections) : false;
};

const Device: FC<Props> = ({ deviceName, deviceData, onDrop }) => {
  const connections = useMemo(() => {
    return deviceData.connections.concat(deviceData.parent?.name || []);
  }, [deviceData.connections, deviceData.parent]);

  const [{ isDragging }, drag] = useDrag({
    item: { type: 'device', deviceData },
    collect: monitor => ({
      isDragging: !!monitor.isDragging(),
    }),
  });

  const [{ isHover }, drop] = useDrop({
    accept: 'device',
    drop: (item: any, monitor) => {
      if (monitor.canDrop()) {
        onDrop(item.deviceData.name, deviceData.name);
        return;
      }
    },
    canDrop: (item: any): boolean => {
      const { deviceData: {Â name } } = item;
      return isValidLink(name, deviceData.name, connections);
    },
    collect: (monitor) => {
      return {
        isHover: monitor.isOver() && monitor.canDrop(),
      };
    },
  });

  const deviceClassName = useMemo(() => {
    if (connections.length === 0)
      return DeviceColor.EMPTY;
    return deviceColorClasses.get(deviceName);
  }, [connections.length, deviceName]);

  return (
    <div className="device-box">
      <div
        ref={mergeRefs([drag, drop])}
        className={`device-name-box ${deviceClassName}${isDragging ? ' is-dragging' : ''}${isHover ? ' is-hover' : ''}`}
      >
        <div>{deviceData.name}</div>
      </div>
      <div className="vertical-bar" />
      <div className="connections-container">
        <div>Connections:</div>
        <div className="connections-boxes">
          {connections.map((connection, idx) => (
              <div
                id={`${connection}${idx}`}
                className={`connection ${deviceColorClasses.get(connection[0])}`}
                key={`${connection}`}
                style={{ cursor: 'pointer' }}
              >
                {connection}
              </div>
            ))}
        </div>
      </div>
      <div className={'trash-icon-container'}>
        <div className="vertical-bar" />
        <div className={'trash-icon'}>
          <FontAwesomeIcon icon={faTrashAlt} />
        </div>
      </div>
    </div>
  );
};

export default Device;
