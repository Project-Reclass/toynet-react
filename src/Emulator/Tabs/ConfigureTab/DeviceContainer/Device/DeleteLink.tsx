import React, { FC } from 'react';
import { useDrop } from 'react-dnd';
import { Icon } from '@chakra-ui/core';
import styled from '@emotion/styled';
import { Flex, Divider } from '@chakra-ui/core';

import { LinkFunc } from './shared';

const HoverableTrashCan = styled.div`
  margin: auto;

  > svg {
    transition: 0.15s opacity ease-in-out;
    opacity: ${({isHover}: {isHover: boolean}) => isHover ? '1' : '0.2'}
  }
`;

interface Props {
  name: string;
  onRemove: LinkFunc;
}

const DeleteLink: FC<Props> = ({ name, onRemove }) => {
  const [{ isHover }, drop] = useDrop({
    accept: 'device',
    canDrop: (item: any) => {
      const { deviceData } = item;
      return deviceData.isLink || deviceData.connections.length === 0;
    },
    drop: (item: any, monitor) => {
      if (monitor.canDrop())
        onRemove(item.deviceData.name, name);
    },
    collect: (monitor) => {
      return {
        isHover: monitor.isOver() && monitor.canDrop(),
      };
    },
  });

  return (
    <Flex marginLeft='auto' ref={drop} data-testid='trash-icon'>
      <Divider orientation='vertical' />
      <HoverableTrashCan isHover={isHover}>
        <Icon name='delete' size='1.2rem' focusable={true} />
      </HoverableTrashCan>
    </Flex>
  );
};

export default DeleteLink;