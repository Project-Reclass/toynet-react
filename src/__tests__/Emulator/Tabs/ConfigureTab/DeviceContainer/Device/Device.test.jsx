import React from 'react';
import { DndProvider } from 'react-dnd';
import renderer, { act } from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import { HTML5Backend } from 'react-dnd-html5-backend';

import Device from 'src/Emulator/Tabs/ConfigureTab/DeviceContainer/Device';

function withDnDProvider(Component) {
  return (props) => (
    <DndProvider backend={HTML5Backend}>
      <Component {...props} />
    </DndProvider>
  )
}

const defaultProps = {
  deviceName: 'Router',
  deviceData: {
    name: 'test',
    connections: [],
  }
};

const TestComponent = withDnDProvider(Device);

const InteractionTestComponent = ({ deviceOne, deviceTwo }) => {
  return (
    <div>
      <TestComponent {...deviceOne} />
      <TestComponent {...deviceTwo} />
    </div>
  )
}

const onDropMock = jest.fn();

const routerConnectionsData = {
  deviceName: 'r1',
  connections: ['h1', 'h2', 'h3']
};

const hostConnections = {
  deviceName: 'h1',
  connections: ['r1', 'r2', 'r3']
};

const switchConnections = {
  deviceName: 's1',
  connections: ['s2', 's3', 's4']
};

const routerDeviceMock = {
  deviceName: 'Router',
  deviceData: {
    name: 'router-one',
    type: 'router',
    connections: [],
  },
  onDrop: onDropMock,
}
const switchDeviceMock = {
  deviceName: 'Switch',
  deviceData: {
    name: 'switch-one',
    type: 'switch',
    connections: [],
  },
  onDrop: onDropMock,
}

const hostDeviceMock = {
  deviceName: 'Host',
  deviceData: {
    name: 'host-one',
    type: 'host',
    connections: [],
  },
  onDrop: onDropMock,
}

describe('The Device', () => {
  beforeEach(() => {
    onDropMock.mockClear();
  });

  it('should match previous snapshots', () => {
    const tree = renderer.create(<TestComponent {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match previous snapshot with router connections', () => {
    const tree = renderer.create(<TestComponent {...defaultProps} deviceData={routerConnectionsData} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match previous snapshots with host connections', () => {
    const tree = renderer.create(<TestComponent {...defaultProps} deviceData={hostConnections} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match previous snapshots with switch connections', () => {
    const tree = renderer.create(<TestComponent {...defaultProps} deviceData={switchConnections} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should allow routers to connect switches', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={routerDeviceMock}
      deviceTwo={switchDeviceMock}
    />);

    const routerDevice = getByText('router-one');
    const switchDevice = getByText('switch-one');

    act(() => {
      fireEvent.dragStart(routerDevice);
      fireEvent.drag(routerDevice);
      fireEvent.dragOver(switchDevice);
      fireEvent.drop(switchDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should allow routers to connect to other routers', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={routerDeviceMock}
      deviceTwo={{...routerDeviceMock, deviceData: { ...routerDeviceMock.deviceData, name: 'router-two' }}}
    />);

    const routerDevice = getByText('router-one');
    const switchDevice = getByText('router-two');

    act(() => {
      fireEvent.dragStart(routerDevice);
      fireEvent.drag(routerDevice);
      fireEvent.dragOver(switchDevice);
      fireEvent.drop(switchDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should not allow routers to connect to hosts', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={routerDeviceMock}
      deviceTwo={hostDeviceMock}
    />);

    const routerDevice = getByText('router-one');
    const hostDevice = getByText('host-one');

    act(() => {
      fireEvent.dragStart(routerDevice);
      fireEvent.drag(routerDevice);
      fireEvent.dragOver(hostDevice);
      fireEvent.drop(hostDevice);
    });

    expect(onDropMock).not.toHaveBeenCalled();
  });

  it('should allow switches to connect to routers', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={switchDeviceMock}
      deviceTwo={routerDeviceMock}
    />);

    const routerDevice = getByText('router-one');
    const switchDevice = getByText('switch-one');

    act(() => {
      fireEvent.dragStart(switchDevice);
      fireEvent.drag(switchDevice);
      fireEvent.dragOver(routerDevice);
      fireEvent.drop(routerDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should allow switches to connect other switches', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={switchDeviceMock}
      deviceTwo={{...switchDeviceMock, deviceData: { ...switchDeviceMock.deviceData, name: 'switch-two' }}}
    />);

    const switchTwoDevice = getByText('switch-two');
    const switchDevice = getByText('switch-one');

    act(() => {
      fireEvent.dragStart(switchDevice);
      fireEvent.drag(switchDevice);
      fireEvent.dragOver(switchTwoDevice);
      fireEvent.drop(switchTwoDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should allow switches to connect to hosts', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={switchDeviceMock}
      deviceTwo={hostDeviceMock}
    />);

    const switchDevice = getByText('switch-one');
    const hostDevice = getByText('host-one');

    act(() => {
      fireEvent.dragStart(switchDevice);
      fireEvent.drag(switchDevice);
      fireEvent.dragOver(hostDevice);
      fireEvent.drop(hostDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should allow hosts to connect to switches', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={hostDeviceMock}
      deviceTwo={switchDeviceMock}
    />);

    const hostDevice = getByText('host-one');
    const switchDevice = getByText('switch-one');

    act(() => {
      fireEvent.dragStart(hostDevice);
      fireEvent.drag(hostDevice);
      fireEvent.dragOver(switchDevice);
      fireEvent.drop(switchDevice);
    });

    expect(onDropMock).toHaveBeenCalled();
  });

  it('should not allow hosts to connect to other hosts', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={hostDeviceMock}
      deviceTwo={{...hostDeviceMock, deviceData: { ...hostDeviceMock.deviceData, name: 'host-two' }}}
    />);

    const hostDevice = getByText('host-one');
    const hostTwoDevice = getByText('host-two');

    act(() => {
      fireEvent.dragStart(hostDevice);
      fireEvent.drag(hostDevice);
      fireEvent.dragOver(hostTwoDevice);
      fireEvent.drop(hostTwoDevice);
    });

    expect(onDropMock).not.toHaveBeenCalled();
  });

  it('should not allow hosts to connect to a router', () => {
    const { getByText } = render(<InteractionTestComponent
      deviceOne={hostDeviceMock}
      deviceTwo={routerDeviceMock}
    />);

    const hostDevice = getByText('host-one');
    const routerDevice = getByText('router-one');

    act(() => {
      fireEvent.dragStart(hostDevice);
      fireEvent.drag(hostDevice);
      fireEvent.dragOver(routerDevice);
      fireEvent.drop(routerDevice);
    });

    expect(onDropMock).not.toHaveBeenCalled();
  });

  // it('should only allow a host to have one connection', () => {
  //   const { getByText } = render(<InteractionTestComponent
  //     deviceOne={{...hostDeviceMock, deviceData: { ...hostDeviceMock.deviceData, connections: ['s1'] }}}
  //     deviceTwo={switchDeviceMock}
  //   />);

  //   const hostDevice = getByText('host-one');
  //   const switchDevice = getByText('switch-one');

  //   act(() => {
  //     fireEvent.dragStart(hostDevice);
  //     fireEvent.drag(hostDevice);
  //     fireEvent.dragOver(switchDevice);
  //     fireEvent.drop(switchDevice);
  //   });

  //   expect(onDropMock).not.toHaveBeenCalled();
  // });
});