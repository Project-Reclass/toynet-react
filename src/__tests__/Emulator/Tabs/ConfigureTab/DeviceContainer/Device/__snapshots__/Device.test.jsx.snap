// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Device should match previous snapshot with router connections 1`] = `
<div
  className="device-box"
>
  <div
    className="device-name-box router-color"
  >
    <div />
  </div>
  <div
    className="vertical-bar"
  />
  <div
    className="connections-container"
  >
    <div>
      Connections:
    </div>
    <div
      className="connections-boxes"
    >
      <div
        className="connection host-color"
        id="h10"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        h1
      </div>
      <div
        className="connection host-color"
        id="h21"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        h2
      </div>
      <div
        className="connection host-color"
        id="h32"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        h3
      </div>
    </div>
  </div>
  <div
    className="trash-icon-container"
  >
    <div
      className="vertical-bar"
    />
    <div
      className="trash-icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 "
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`The Device should match previous snapshots 1`] = `
<div
  className="device-box"
>
  <div
    className="device-name-box empty-color"
  >
    <div>
      test
    </div>
  </div>
  <div
    className="vertical-bar"
  />
  <div
    className="connections-container"
  >
    <div>
      Connections:
    </div>
    <div
      className="connections-boxes"
    />
  </div>
  <div
    className="trash-icon-container"
  >
    <div
      className="vertical-bar"
    />
    <div
      className="trash-icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 "
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`The Device should match previous snapshots with host connections 1`] = `
<div
  className="device-box"
>
  <div
    className="device-name-box router-color"
  >
    <div />
  </div>
  <div
    className="vertical-bar"
  />
  <div
    className="connections-container"
  >
    <div>
      Connections:
    </div>
    <div
      className="connections-boxes"
    >
      <div
        className="connection router-color"
        id="r10"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        r1
      </div>
      <div
        className="connection router-color"
        id="r21"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        r2
      </div>
      <div
        className="connection router-color"
        id="r32"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        r3
      </div>
    </div>
  </div>
  <div
    className="trash-icon-container"
  >
    <div
      className="vertical-bar"
    />
    <div
      className="trash-icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 "
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`The Device should match previous snapshots with switch connections 1`] = `
<div
  className="device-box"
>
  <div
    className="device-name-box router-color"
  >
    <div />
  </div>
  <div
    className="vertical-bar"
  />
  <div
    className="connections-container"
  >
    <div>
      Connections:
    </div>
    <div
      className="connections-boxes"
    >
      <div
        className="connection switch-color"
        id="s20"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        s2
      </div>
      <div
        className="connection switch-color"
        id="s31"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        s3
      </div>
      <div
        className="connection switch-color"
        id="s42"
        style={
          Object {
            "cursor": "pointer",
          }
        }
      >
        s4
      </div>
    </div>
  </div>
  <div
    className="trash-icon-container"
  >
    <div
      className="vertical-bar"
    />
    <div
      className="trash-icon"
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-trash-alt fa-w-14 "
        data-icon="trash-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
